opencog/query/PatternMatch.cc

```diff
@@ -328,13 +329,33 @@ bool PatternMatch::recursive_virtual(PatternMatchCallback& cb,
  * to indicate the bindings of the variables, and (optionally) limit
  * the types of acceptable groundings for the variables.
  */
-bool BindLink::imply(PatternMatchCallback& pmc, bool check_conn)
+bool BindLink::imply(PatternMatchCallback& pmc, AtomSpace* as, bool check_conn)
 {
    if (check_conn and 0 == _virtual.size() and 1 < _components.size())
 		throw InvalidParamException(TRACE_INFO,
 			"BindLink consists of multiple disconnected components!");
 			// PatternLink::check_connectivity(_comps);
 
+	// Make sure that the user did not pass in bogus clauses
+	// in the queried atomspace.
+	// Make sure that every clause contains at least one variable.
+	// The presence of constant clauses will mess up the current
+	// pattern matcher.  Constant clauses are "trivial" to match,
+	// and so its pointless to even send them through the system.
+	bool bogus = remove_constants(_varlist.varset, _pat, _components, as);
```

Comment:
well, you've moved this code out of the compiled code path, and into the executed code path.  Now, instead of checking for constants just once, when the pattern is created, you are checking them every time that the pattern is executed -- and since a pattern can be re-used thousands of times, that's a thousand-fold impact on performance.

Reply:
Yes, I know, but you *can't* check whether a constant is actually present in the to-be-queried atomspace at compile time. Besides, seriously, how much performance overhead is it gonna have, +0.001%? it's still insignificant compared to the search itself.

Reply:
Eh? when I last measured, the search was fast, and the building of the bindlinks were slow. There's a lot of overhead and calculation to construct a bind-link (by design).  Of course, it's data-dependent, but the actual execution of a bindlink should be fast.  Normally, there's not much to do.